system-addon/lib/HighlightsFeed.jsm

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {classes: Cc, interfaces: Ci, utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/NewTabUtils.jsm");
+Cu.importGlobalProperties(["fetch"]);
+
+const {actionTypes: at} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+
+const {shortURL} = Cu.import("resource://activity-stream/lib/ShortURL.jsm", {});
+const {SectionsManager} = Cu.import("resource://activity-stream/lib/SectionsManager.jsm", {});
+
+XPCOMUtils.defineLazyModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const HIGHLIGHTS_MAX_LENGTH = 9;
+const HIGHLIGHTS_UPDATE_TIME = 15 * 60 * 1000; // 15 minutes
+const SECTION_ID = "highlights";
+
+this.HighlightsFeed = class HighlightsFeed {
+  constructor() {
+    this.highlightsLastUpdated = 0;
+    this.highlights = [];
+  }
+
+  init() {
+    SectionsManager.onceInitialized(this.postInit.bind(this));
+  }
+
+  postInit() {
+    SectionsManager.enableSection(SECTION_ID);
+    this.fetchHighlights();
+  }
+
+  uninit() {
+    SectionsManager.disableSection(SECTION_ID);
+  }
+
+  async fetchHighlights(broadcast = false) {
+    this.highlights = await NewTabUtils.activityStreamLinks.getHighlights();
+    for (let highlight of this.highlights) {
+      highlight.hostname = shortURL(Object.assign({}, highlight, {url: highlight.url}));
+      highlight.image = highlight.preview_image_url;
+    }
+
+    // Remove any Highlights that are in Top Sites already
+    const topsites = this.store.getState().TopSites.rows.map(site => site.url);
+    this.highlights = this.highlights.filter(site => topsites.indexOf(site.url) === -1);
+
+    SectionsManager.updateSection(SECTION_ID, {rows: this.highlights}, this.highlightsLastUpdated === 0 || broadcast);
+    this.highlightsLastUpdated = Date.now();
+  }
+
+  onAction(action) {
+    switch (action.type) {
+      case at.INIT:
+        this.init();
+        break;
+      case at.NEW_TAB_LOAD:
+        if (this.highlights.length < HIGHLIGHTS_MAX_LENGTH) {
+          // If we haven't filled the highlights grid yet, fetch again.
+          this.fetchHighlights(true);
+        } else if (Date.now() - this.highlightsLastUpdated >= HIGHLIGHTS_UPDATE_TIME) {
+          // If the last time we refreshed the data is greater than 15 minutes, fetch again.
+          this.fetchHighlights(false);
+        }
+        break;
+      case at.MIGRATION_COMPLETED:
+      case at.PLACES_HISTORY_CLEARED:
+      case at.PLACES_LINK_DELETED:
+      case at.PLACES_LINK_BLOCKED:
+        this.fetchHighlights(true);
+        break;
+      case at.PLACES_BOOKMARK_ADDED:
+      case at.PLACES_BOOKMARK_REMOVED:
+        this.fetchHighlights(false);
```

Comment:
I see that we avoid updating the UI here. But the context menu doesn't update either. Perhaps the Section Reducer isn't working correctly?